### DISCUSSION TITLE
Documentation Roadmap

### DISCUSSION THREAD

**Discussion by HalilFocic** [maxGraph/maxGraph#595] [Development]
Hello. I would like to contribute to documentation since I dont think it's very organized. Might be personal but I had really hard time trying to figure stuff since my project at work required a lot of stuff that doesn't come out of the box and you had to manually do the digging. My main source or inspiration was the full featured editor by jgraph. 

I don't think the mxGraph documentation is really any better but we can for sure take a lot of stuff from them to power our Docusaurus docs. 
Storybook provides really good examples but its quite hard to figure out where to look to find the source code you need. 

At current moment, I don't know if the priority is on documentation or fixing bugs/adding features. I am open to everything but I think we are losing a lot of users due to bad documentation.

**Comment by HalilFocic** [comment]
@tbouffard If you think we need to prioritize on featues, just make feature issues and assign them to me, I'll handle it but what I see from Issue list is that we got a really small number of bugs in core, most of them are storybook related. I dont think fixing manhattan connector is more important than docs since really few people will use the connector and a lot of people need proper docs.

**Reply by tbouffard** [reply]
Thanks a lot for bringing this up!

Just to give a bit of context, no clear priorities have been set for topics in the project right now. I've been the only maintainer of maxGraph for quite some time now, after the two main contributors (who did a lot of work on the mxGraph migration) had to step back. Since then, I've been picking issues based on my own needs and the small amount of feedback we've gotten, as the active user base isn't very big.



I totally agree that new users face challenges when getting started with maxGraph. Improving the documentation is key to fixing this. The mxGraph documentation is a good starting point, which is why I started integrating it into maxGraph. But there's still more to do—like adding tutorial content and getting rid of old mxGraph references.



You're right, though—the mxGraph docs alone aren't enough. There are still a lot of gaps, and there aren't dedicated pages for some features. That's why I made the Perimeters page and started working on the Codec/Serialization page (see [PR #524](https://github.com/maxGraph/maxGraph/pull/524)).



I think focusing on documentation would be a big improvement. A good starting point would be to finish the migration of the mxGraph docs ([Issue #345](https://github.com/maxGraph/maxGraph/issues/345)) and reorganize things so that tutorials and guides flow better. We could also think about merging the manual and tutorial into one resource, since the difference between them has always been a bit unclear to me. Additionally, we could figure out which topics or features need their own dedicated pages to complement the main tutorial or manual.

We could also link relevant stories in the docs to show those topics in action.



The examples also need some rework. We don't have enough examples that focus on specific features. A lot of the current examples mix multiple features, which makes it hard to know what they're about just from the title. That's why I recently started adding more focused examples, like in [PR #576](https://github.com/maxGraph/maxGraph/pull/576) and [PR #586](https://github.com/maxGraph/maxGraph/pull/586).

As you mentioned, the more complex examples, especially the ones about extensions, are still in the GraphEditor and haven't been migrated yet. We haven't decided if we will migrate or remove them, and this seems like a separate topic we could discuss in its own thread.



Let me know what you think.

**Reply by HalilFocic** [reply]
I am not quite sure what are the best paths to take here. I can tackle the migration from mxGraph to our docs and once that is finished we can maybe discuss where to take it from here. I would love to see more links on the left side. I dont think a lot of people will connect the dots when they go to "API" page. It has everything but going through it is difficult. I think tutorial should take you from 0 to graph with some features and basic understanding of maxgraph. Current tutorial doesn't do it. So lets migrate the mxGraph docs first and them reorganize and add sections on the left. 

You can assign any docs related issue to me and I will handle it when I have the time.

**Reply by tbouffard** [reply]
Hi @HalilFocic , and Happy New Year! 🎉  

To get started, I suggest we move the former mxGraph contents into the documentation and adapt them to the new API. To make the implementation easier, I’ve split [issue #345](https://github.com/maxGraph/maxGraph/issues/345) into smaller tasks. Currently, I’m working on [issue #613](https://github.com/maxGraph/maxGraph/issues/613).  

In parallel, we could discuss how to structure the documentation effectively. It’s important to define the types of guides, tutorials, and detailed resources we want to offer. A good starting point might be [Diátaxis](https://diataxis.fr/), which provides a helpful framework for organizing documentation.  We could also follow the advice provided in https://github.blog/developer-skills/documentation-done-right-a-developers-guide/.

Here’s a non-exhaustive list of elements that could add value to our users:  

### General Suggestions
- Add links to existing examples and/or stories on all pages to illustrate documented concepts.  
  Many users have mentioned that finding examples is challenging (some even think we’re missing key examples). Improving this could greatly enhance the user experience.  
  For reference, I recently did this in the [Codecs page update](https://github.com/maxGraph/maxGraph/pull/524).  

### Specific Topics

- **Codecs**
  - add info about the conversion of boolean and numeric attribute, see https://github.com/maxGraph/maxGraph/issues/605
- **Connectors/Edge Styles**  
  - Page creation with introduction and basic information:  #791
  - How to use (including loops), configure, and create custom elements.  
  - Custom implementation: see the Wires story
  - Add a link to the Perimeter page about the "orthogonal projection" for connector considered as orthogonal (explain what it is)
  - Add a link to the EdgeHandler page (the EdgeHandler implementation depends on the type of connector)
- **Edge Markers**  
  - Many additional markers are available in draw.io. For example, I’ve used [this one](https://github.com/jgraph/drawio/blob/f539f1ff362e76127dcc7e68b5a9d83dd7d4965c/src/main/webapp/js/mxgraph/Shapes.js#L2796) in a project.  
  - We could add an example of a custom marker (e.g., a small filled square).  
- **EdgeHandler**  
  - Focus on options like enabling bends and other customizations.  
  - Provide a new story demonstrating edge styles and edge handlers (use Storybook controls/args)
- **Event** (internal event, synchronous mechanism) EventSource listener, register listener on the right object, how to know what is the listening object, .... see #1053
- **Export**  (find a better name)
  - SVG export. See https://github.com/maxGraph/maxGraph/discussions/234#discussioncomment-6848935, #604 and #761
  - ImageExport and XmlCanvas2D: serialize the view, in mxGraph, the XML result of the export was sent to a backend service that created PNG, JPG, PDF, ... representation of the Graph
- **Guide**
  - customization and configuration
  - stories showing the feature: at least in DragSource
  - stories extending the default behavior
- **Graph** see #781
- **Images**  
  - some provided in the npm package, need to copy them in your application (mxGraph legacy)
  - how to configure path to images
- **Image Bundle**: see #1050   
- **Keyboard**  
- **Labels**  
  - Bounds and positions 
  - `htmlLabels`, `foreignObject`, or `SVG` text.
  - wrap label into bounds: require htmlLabels. Include wrapping examples, see the Wrapping story.  
- **Languages/Translations**  
  - need to copy them in your application (mxGraph legacy)
  -  how to configure path to resources
- **Layout**
  - ParallelEdgeLayout: see https://github.com/maxGraph/maxGraph/issues/371#issuecomment-2606450554
- **Logs**
  - Available loggers: default implementation (no logs), console, Maxlog (require CSS and show function call to display)
  - How to use a specific logger implementation
  - How to implement a Custom logger: implement dedicated interface + usage
- **Navigation**  
  - Pan, fit, scroll, translate, zoom, gestures (with limitations), mouse interactions
  - Link to the ZoomAndFit story and ts-example (since 0.17.0).
- **Overlays** 
- **Perimeters** (update) 
  - entryPerimeter/exitPerimeter: example see https://stackoverflow.com/a/62127707/14299521
  - spacing: sourcePerimeterSpacing, targetPerimeterSpacing, perimeterSpacing
  - link with entryX, entryY, entryDx, entryDy
  - special use case, source/target is an edge. This is not documented but this works, see https://github.com/process-analytics/bpmn-visualization-js/issues/136, use case: [BPMN Association](https://creately.com/guides/bpmn-symbols/#3-artifacts)
  - Add Link to Edgestyle in paragraph about orthogonal projection 
- **Plugins**  update existing page
  - lifecycle
  - list all builtins plugin and show the plugins considered as "default plugins" (in a table)
  - explain the purpose of each plugin, give example, link to existing page (for example ImageBundle introduced in #1050), link to stories
  - how to configure: retrieve it and call its methods/properties
  - link to the Graph page
- **Shapes + Custom Shapes**  
  - How to implement and register them. 
  - draw.io provides a lot of shapes as stencils and with JavaScript implementation
- **Stencils**  
  - Links to the existing stories
  - Links to draw.io stencil definitions that can be reused in maxGraph. 
  - draw.io has the concept of stencil library: mention it
  - draw.io documentation:
    - https://www.drawio.com/doc/faq/custom-shapes
    - https://www.drawio.com/doc/faq/shape-complex-create-edit 
- **Style** curved/rounded  
  - Edges:
    - differences between curved (link to the `EdgeCurvedAndRounded` story updated in 0.17.0) and rounded
    - arcSize
  - Shapes:  
    - arcSize and relative/absolute size
   - rectangle rounded factor
- **Stylesheet**  
   - can be loaded with a codec
   - draw.io provide xml resources to be loaded
- **Tree Shaking**  
  - #781  
- **UI Components**: MaxForm, MaxLog, MaxToolBar, MaxWindow, ... require CSS file
- **VertexHandler**
  - Options and customizations.  
- **General topics**
  - Alternate style 
  - Constraints
  - Floating point
  - Port: another shape can be used to determine where the terminal point are put. Use case: inner elements, 
fine for editor, not always accurate for visualization when the terminal point coordinates are provided
  - Tutorial and Manual: content has been migrated from the mxGraph documentation. To review has they include duplicated information. The content should be reviewed in the light of what exists on the “usage” pages.
- **Adding automatic tests involving maxGraph within an application** (as a guide)
  - setup jest: #790
  - other test frameworks
- Special Use. Not the target of maxgraph, nor really supported, but why not. Could be in a FAQ
  - node js: see the node.js example and #761
  - directly in browser page: see #1051 (repro page) and #462


What do you think about these ideas?  
Feel free to share your thoughts and proposals.

**Reply by HalilFocic** [reply]
@tbouffard Happy new year to you as well. 

I would have to look up about Diataxis since I am unfamiliar with it. I dont know the every topic you listed here but I can start with the ones Im atleast a bit familiar with.

Also I've noticed that in navigation you say we are limited. I would love to add support for some features and fix bugs if they are present. Thats something I really enjoy, making the software work better. It would also be a plus for our package to work on mobile out of the box.

I'll create PRs with some outline for the topics I am familiar with when I get time to do so.

Cheers.

**Reply by tbouffard** [reply]
Hi @HalilFocic 
I've made some progress in integrating mxGraph content into the maxGraph documentation. I've also started to reorganize the pages with this content.
I've also suggested #625 to better highlight the improvements to the getting-started page you've made in recent weeks. I'd like to hear what you think.

**Reply by HalilFocic** [reply]
> Hi @HalilFocic 
> 
> I've made some progress in integrating mxGraph content into the maxGraph documentation. I've also started to reorganize the pages with this content.
> 
> I've also suggested #625 to better highlight the improvements to the getting-started page you've made in recent weeks. I'd like to hear what you think.

Getting started looks good to me. Nice
